### PR TITLE
Add converter aten::arange

### PR DESCRIPTION
**Pull request by inocsin**
# Description

Please merge this PR after https://github.com/NVIDIA/TRTorch/pull/324

Fixes https://github.com/NVIDIA/TRTorch/issues/105

## Type of change

Add converter aten::arange

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by narendasan — COMMENTED** [review]
`arange` just produces a tensor from a start and end value right? Is there a reason this could not be an evaluator that creates a static tensor? What cases would this be abstract at compile time?

**Comment by inocsin** [comment]
> `arange` just produces a tensor from a start and end value right? Is there a reason this could not be an evaluator that creates a static tensor? What cases would this be abstract at compile time?

Can evaluator return a tensor?

**Comment by narendasan** [comment]
Evaluators can return torch tensors

**Comment by inocsin** [comment]
> Evaluators can return torch tensors
```
  std::unordered_map<const torch::jit::Value*, nvinfer1::ITensor*> value_tensor_map;
  std::unordered_map<const torch::jit::Value*, torch::jit::IValue> evaluated_value_map;
```
evalutor can only return torch::jit::IValue type right? So I use vector instead of at::Tensor, but this code doesn't work for me
```cpp
        .evaluator({c10::Symbol::fromQualString("aten::arange"),
                    [](const torch::jit::Node* n, kwargs& args) -> c10::optional<torch::jit::IValue> {
                      int end_scalar = 0;
                      if (args.at(n->input(0)).IValue()->isInt()) {
                        end_scalar = args.at(n->input(0)).unwrapToInt();
                      } else if (args.at(n->input(0)).IValue()->isDouble()) {
                        end_scalar = ceil(args.at(n->input(0)).unwrapToScalar().to<float>());
                      }
                      std::vector<int> arr;
                      for (int i = 0; i < end_scalar; i++) {
                        arr.push_back(i);
                      }
                      return arr;
                    },
                    EvalOptions().validSchemas({
                        R"SIG(aten::arange(Scalar end, *, int? dtype=None, int? layout=None,
                            Device? device=None, bool? pin_memory=None) -> (Tensor))SIG",
                    })});
```

```cpp
TEST(Evaluators, ATenArangeIntEvaluatesCorrectly) {
  const auto graph = R"IR(
      graph():
        %0 : int = prim::Constant[value=51]()
        %1 : None = prim::Constant()
        %2 : Tensor = aten::arange(%0, %1, %1, %1, %1)
        return (%2))IR";

  auto g = std::make_shared<torch::jit::Graph>();
  torch::jit::parseIR(graph, &*g);

  auto jit_results = trtorch::tests::util::EvaluateGraphJIT(g, {});
  auto trt_results = trtorch::tests::util::EvaluateGraph(g->block(), {});

  ASSERT_TRUE(jit_results[0] == trt_results[0]);
}
```

Here is error log
```
unknown file: Failure
C++ exception with description "Type int could not be converted to any of the known types.
Exception raised from call at bazel-out/k8-fastbuild/bin/external/libtorch/_virtual_includes/ATen/ATen/core/jit_type.h:1703 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x69 (0x7f78b895fb89 in /home/vincentzh/Projects/TRTorch/bazel-TRTorch/external/libtorch/lib/libc10.so)
frame #1: c10::detail::getTypePtr_<int>::call() + 0x109 (0x7f793f4567b0 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #2: std::shared_ptr<c10::Type> c10::getTypePtr<int>() + 0x18 (0x7f793f45623b in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #3: c10::List<int>::List() + 0x19 (0x7f793f455b1d in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #4: c10::IValue::IValue<int, (decltype(nullptr))0>(std::vector<int, std::allocator<int> > const&) + 0x1d (0x7f793f4550f9 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #5: c10::optional<c10::IValue>::optional<std::vector<int, std::allocator<int> >&, false>(std::vector<int, std::allocator<int> >&) + 0x33 (0x7f793f453f1f in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #6: <unknown function> + 0x9825a (0x7f793f44225a in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #7: <unknown function> + 0x9c4e7 (0x7f793f4464e7 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #8: std::function<c10::optional<c10::IValue> (torch::jit::Node const*, std::map<torch::jit::Value const*, trtorch::core::conversion::Var, std::less<torch::jit::Value const*>, std::allocator<std::pair<torch::jit::Value const* const, trtorch::core::conversion::Var> > >&)>::operator()(torch::jit::Node const*, std::map<torch::jit::Value const*, trtorch::core::conversion::Var, std::less<torch::jit::Value const*>, std::allocator<std::pair<torch::jit::Value const* const, trtorch::core::conversion::Var> > >&) const + 0x66 (0x7f793f419be8 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #9: trtorch::core::conversion::evaluators::EvalNode(torch::jit::Node const*, std::map<torch::jit::Value const*, trtorch::core::conversion::Var, std::less<torch::jit::Value const*>, std::allocator<std::pair<torch::jit::Value const* const, trtorch::core::conversion::Var> > >&) + 0x57 (0x7f793f416855 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Sevaluators_Slibevaluators.so)
frame #10: trtorch::core::conversion::EvaluateNode(trtorch::core::conversion::ConversionCtx*, torch::jit::Node const*, int, int) + 0x6b1 (0x7f7941c50b1e in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libcore_Sconversion_Slibconversion.so)
frame #11: trtorch::tests::util::EvaluateGraph(torch::jit::Block const*, std::vector<c10::IValue, std::allocator<c10::IValue> >) + 0x600 (0x7f7941d6890b in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libtests_Sutil_Slibutil.so)
frame #12: <unknown function> + 0x1fe6d (0x5600e71dce6d in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/test_aten_evaluators.runfiles/TRTorch/tests/core/conversion/evaluators/test_aten_evaluators)
frame #13: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x65 (0x7f7922e9ca72 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #14: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x5a (0x7f7922e9825b in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #15: testing::Test::Run() + 0xee (0x7f7922e82648 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #16: testing::TestInfo::Run() + 0x10f (0x7f7922e82fa9 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #17: testing::TestSuite::Run() + 0x12c (0x7f7922e83680 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #18: testing::internal::UnitTestImpl::RunAllTests() + 0x41a (0x7f7922e8ef6a in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #19: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x65 (0x7f7922e9daf5 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #20: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x5a (0x7f7922e990ff in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #21: testing::UnitTest::Run() + 0xc9 (0x7f7922e8d8dd in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest.so)
frame #22: RUN_ALL_TESTS() + 0x11 (0x7f7941ba1aa8 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest_Umain.so)
frame #23: main + 0x3d (0x7f7941ba1a37 in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/../../../../_solib_k8/libexternal_Sgoogletest_Slibgtest_Umain.so)
frame #24: __libc_start_main + 0xe7 (0x7f78b79b1b97 in /lib/x86_64-linux-gnu/libc.so.6)
frame #25: <unknown function> + 0x1f4ba (0x5600e71dc4ba in /home/vincentzh/.cache/bazel/_bazel_vincentzh/e4aa060bba3b1641e3659fc2f09887e8/execroot/TRTorch/bazel-out/k8-fastbuild/bin/tests/core/conversion/evaluators/test_aten_evaluators.runfiles/TRTorch/tests/core/conversion/evaluators/test_aten_evaluators)
" thrown in the test body.
[  FAILED  ] Evaluators.ATenArangeIntEvaluatesCorrectly (24 ms)
```

**Comment by peri044** [comment]
I think you can directly use torch::arange internally. For example I tried the following 
```
int end_scalar = 0;
if (args.at(n->input(0)).IValue()->isInt()) {
  end_scalar = args.at(n->input(0)).unwrapToInt();
} else if (args.at(n->input(0)).IValue()->isDouble()) {
  end_scalar = ceil(args.at(n->input(0)).unwrapToScalar().to<float>());
}
// std::vector<int> arr;
// for (int i = 0; i < end_scalar; i++) {
//   arr.push_back(i);
// }
LOG_DEBUG("===================Value of arange: " << torch::arange(end_scalar));
return torch::arange(end_scalar);
```
if we have start, end, step, we can use torch::arange(start, end, step) which returns an at::Tensor
and I compared the outputs in the test case using 
```
bool checkRtol(const at::Tensor& diff, const std::vector<at::Tensor> inputs) {
  double maxValue = 0.0;
  for (auto& tensor : inputs) {
    maxValue = fmax(tensor.abs().max().item<float>(), maxValue);
  }
  std::cout << "Max Difference: " << diff.abs().max().item<float>() << std::endl;
  return diff.abs().max().item<float>() <= 2e-5 * maxValue;
}

bool almostEqual(const at::Tensor& a, const at::Tensor& b) {
  auto a_float = a.toType(at::kFloat);
  auto b_float = b.toType(at::kFloat);
  return checkRtol(a_float - b_float, {a_float, b_float});
}
```
instead of ASSERT_TRUE(jit_results[0] == trt_results[0]), I used `almostEqual(jit_results[0].toTensor(),  trt_results[0].toTensor());`
This seems to be working using the testcase.

**Comment by inocsin** [comment]
pls review the code @peri044

File: core/runtime/register_trt_op.cpp [hunk 1]
```diff
@@ -48,7 +48,16 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
...
+  LOG_DEBUG("Initializing CUDA stream");
+  c10::cuda::CUDAStream stream = at::cuda::getStreamFromPool();
+  if (inputs.size() > 0) {
+    stream = c10::cuda::getCurrentCUDAStream(inputs[0].device().index());
+  } else if (outputs.size() > 0) {
+    stream = c10::cuda::getCurrentCUDAStream(outputs[0].device().index());
+  } else {
+    TRTORCH_THROW_ERROR("Invalid graph. No inputs and outputs found.");
+  }
+  LOG_DEBUG("Executing TensorRT engine");
```
Comment (peri044) [inline core/runtime/register_trt_op.cpp:60]: I think I suggested these changes when arange was expressed as a converter. We can remove this now that this is an evaluator. We had a discussion and these changes were deemed not required as we most likely won't see graphs with no inputs. You can remove this piece of code

File: core/conversion/evaluators/aten.cpp [hunk 2]
```diff
@@ -465,7 +465,21 @@ auto aten_registrations TRTORCH_UNUSED =
                       LOG_WARNING("Warning from TorchScript: " << *warning);
                       return {};
                     },
-                    EvalOptions()});
+                    EvalOptions()})
+        .evaluator({c10::Symbol::fromQualString("aten::arange"),
+                    [](const torch::jit::Node* n, kwargs& args) -> c10::optional<torch::jit::IValue> {
+                      int end_scalar = 0;
+                      if (args.at(n->input(0)).IValue()->isInt()) {
```
Comment (peri044) [inline core/conversion/evaluators/aten.cpp:472]: what about the cases where torch arange takes in `start` and `step` values? Is that handled or am I missing something here ?

**Review by peri044 — CHANGES_REQUESTED** [review]

File: core/conversion/evaluators/aten.cpp (hunk 2 above)
Comment (inocsin) [reply core/conversion/evaluators/aten.cpp:472]: ok, I will implement that part too

File: core/conversion/evaluators/aten.cpp [hunk 3]
```diff
@@ -467,7 +467,63 @@ auto aten_registrations TRTORCH_UNUSED =
                       LOG_WARNING("Warning from TorchScript: " << *warning);
                       return {};
                     },
-                    EvalOptions()});
+                    EvalOptions()})
+        .evaluator({c10::Symbol::fromQualString("aten::arange"),
+                    [](const torch::jit::Node* n, kwargs& args) -> c10::optional<torch::jit::IValue> {
+                      // int end_scalar = 0;
+                      // auto end_scalar = ceil(args.at(n->input(0)).unwrapToScalar());
```
Comment (peri044) [inline core/conversion/evaluators/aten.cpp:474]: You can remove these lines.

File: core/conversion/evaluators/aten.cpp [hunk 4]
```diff
@@ -467,7 +467,63 @@ auto aten_registrations TRTORCH_UNUSED =
...
+                        if (args.at(n->input(i)).IValue()->isScalar()) {
+                          scalar_count += 1;
+                        }
+                      }
+                      if (scalar_count == 1) {
+                        if (args.at(n->input(0)).IValue()->isInt()) {
+                          int end_scalar = args.at(n->input(0)).unwrapToInt();
+                          return torch::arange(end_scalar);
+                        } else if (args.at(n->input(0)).IValue()->isDouble()) {
+                          float end_scalar = ceil(args.at(n->input(0)).unwrapToScalar().to<float>());
```
Comment (peri044) [inline core/conversion/evaluators/aten.cpp:487]: ceil might not be needed here right ?

**Review by peri044 — COMMENTED** [review]
LGTM. Minor comments to be addressed

File: core/conversion/evaluators/aten.cpp (hunk 4 above)
Comment (inocsin) [reply core/conversion/evaluators/aten.cpp]: updated

**Review by peri044 — APPROVED** [review]

**Review by narendasan — APPROVED** [review]
LGTM

**Pull request merged**